### DISCUSSION TITLE
Feature Request: PixelBrush<TPixel>

### DISCUSSION THREAD

**Discussion by calrsom** [SixLabors/ImageSharp.Drawing#165] [Ideas] [answered]
I'm trying to use ImageSharp to render polygon data to raster data, where the underlying pixel values represent data values rather than colors.  At first the ability to define a custom `TPixel` implementation seemed like the perfect match.  Digging deeper, I see that values are converted to single floating point values under the hood; this still works for me.  However, unfortunately the drawing library uses a 16-bit int for the `Color `class.  I can understand this choice for actual colors, but it is limiting for representing other data.

I'm looking for an implementation of the SolidBrush that takes a `TPixel` value rather than a `Color`.  That sidesteps the `Color` issue entirely.  You could call this a "PixelBrush" or "ValueBrush".  Looking at the source, it seems trivial to add this, since the `SolidBrush` converts to a `TPixel` internally.  I'd implement this in my own code base, except `BrushApplicator` has a protected constructor, which means that `IBrush` is not really meant to be implemented (I suppose except for wrapping an existing brush).  I can respect that decision being related to the library being in "beta" status and the API not being locked down.

I could probably put together a PR pretty quick by tweaking the `SolidBrush` class as described.

**Comment by tocsoft** [comment, marked as answer]
This idea sounds fine to me, from an API/implementation POV I would suggest creating a new class `SolidBush<TPixel>` whoes `CreateApplicator<T>` throws if the pixels do not match....alternatively we could fast path with the same pixel type, otherwise we could then call `new Color(TBrushPixel.ToRgba32()).ToPixel<TImagePixel>()` (or equivalent) to translate between pixel types.

Both versions of `SolidBrush` (generic and not) would then both use the same internal `SolidBrushApplicator<T>` (it would just need to be made internal rather than private.)

However saying all that I would like to add you can actually create a custom `BrushApplicator<T>` in your own codebase, the protected constructor just mean you will need to call it from your derived types constructor...ie you can just copy the SolidBrushApplicator implementation into your own code and it should compile fine..

**Reply by antonfirsov** [reply]
> the protected constructor just mean you will need to call it from your derived types constructor...ie you can just copy the SolidBrushApplicator implementation into your own code and it should compile fine..

Exactly. The API has been opened up in #149 to support use cases like this. That change should be sufficient to implement a custom brush that solves this problem.

**Reply by calrsom** [reply]
Thanks for the quick reply!

"you can just copy the SolidBrushApplicator implementation into your own code and it should compile fine"
I tried just that, but I see now that the released nuget package doesn't have the recent change from #149.  I'll grab beta 13.8 and see where I get.

I could try to implement a PR as described as well.  I think I understand what you're talking about with moving the `SolidBrushApplicator`.

**Reply by calrsom** [reply]
So I got pretty close with copying `SolidBrush` into my codebase, but ran into two snags:
- `ThreadLocalBlenderBuffers` is marked internal, so I had to copy that class too
- `BrushApplicator.Blender` is marked internal, so I had to create and initialize my own private blender (this seems like a bug if the class is intended to be extended)

These are both able to be worked around, so I think I can get where I want to from here.

**Reply by calrsom** [reply]
I submitted a PR with a `SolidBush<TPixel>` that I think matches what you were looking for: #166

**Reply by JimBobSquarePants** [reply]
I just approved that PR (_since the implementation is good_) but haven't merged yet as I would like further discussion around this point. 

Doing the change for one brush doesn't solve the issue for other brush types nor any future types. 

We should either:
- Fix the extensibility issues.
- Fix the precision issues.

What are everyone's thoughts?

**Reply by antonfirsov** [reply]
> Fix the extensibility issues.

This would either need an abstract color type that can represent arbitrary precision, or re-adding pixel specific API overloads: `Fill<TPixel>(this IImageProcessingContext<TPixel> source, TPixel color)`. None of those seem to worth it.

I'm stuck in analysis paralysis on whether we need #166 in-box, it serves a **very specific** use case, and might be confusing to other users. Maybe we should add it to [SixLabors/Samples](https://github.com/SixLabors/Samples) instead?

**Reply by JimBobSquarePants** [reply]
I don't really think it serves great use in samples because in order to implement it you still have to copy.

- `ThreadLocalBlenderBuffers`
- `BrushApplicator.Blender`

As far as I'm aware their the only limiting factors to users implementing their own pixel specific instances.

**Reply by calrsom** [reply]
The customizable pixel format seems like a headline feature of ImageSharp.  It would be nice to have some way to support drawing with custom pixels.  At the same time, I see why the Color class is designed the way it is.

**Reply by antonfirsov** [reply]
It's very rare though, that a pixel format goes beyond `Rgba64` precision levels. Applying a `SolidBrush<TPixel>` on non-generic images that conditionally works as a `TPixel` brush is not a straightforward API that's hard to communicate.

The most general way to do this is to add generic drawing overloads, like:
```C#
public static IImageProcessingContext<TPixel> Fill<TPixel>(this IImageProcessingContext<TPixel> ctx, IPath polygon, TPixel color);
```
However if we expand this to all brushes/overloads, that would unnecessarily explode the API surface for vast majority of our users.

I'm can't decide what is the most elegant, or better -- the least worst solution here.

**Reply by tocsoft** [reply]
Because we don't deem the `Color` type to be a performance critical code path could we not smuggle an initializing pixel inside and if the exact pixel type is then later requested back from it then it can return the original value rather then doing any conversion at all. That way, for those people using higher precision types, it should generally just work as expected without the surprise that it looses precision.

**Reply by calrsom** [reply]
Having a TPixel-based Color does seem like the most elegant way from the API side to support this use case, but I'm not sure what that would mean for the implementation.

**Reply by antonfirsov** [reply]
We can carry around an optional boxed pixel instance:

```diff
public class Color
{
    private readonly Rgba64 data;
+   private IPixel boxedHighPrecisionPixel; // null most of the cases
}
```

We still need a trick to decide if a pixel should be boxed instead of being converted to `Rgba64`. Maybe we can add the logic to `PixelOperations<T>`:

```diff
public class PixelOperations<TPixel>
{
    // Heuristic default logic, a custom implementation can override it if necessary:
+   public virtual bool IsHighPrecisionPixel => Unsafe.SizeOf<TPixel> > sizeof(Rgba64);
}
```

@tocsoft @JimBobSquarePants if we are fine with this idea, we should add a tracking issue to https://github.com/SixLabors/ImageSharp and close #166.

**Reply by JimBobSquarePants** [reply]
Couldn't we change the backing `TPixel` type to `RgbaVector`? and add a `Color.FromPixel<TPixel>()` that uses `[From/To]ScaledVector4` to convert? It's already got a `FromHex` method in the type.

**Reply by tocsoft** [reply]
I would still suggest we carry the original boxed pixel as that will always 100% work no matter the pixel type, there is still a potential loss of precision with any sort of conversion were-as if we retain the original pixel it allows us to guarantee no precision loss if you use the same pixel type for initialising a `Color` and the pixel type for the `Image` itself.

**Reply by antonfirsov** [reply]
`Vector4` also brings floating point flakiness, it's better to keep the backing storage integer-based.

From OP:
> [...] where the underlying pixel values represent data values rather than colors.

@calrsom I wonder what is your use case, and what is the data you store in a pixel?

**Reply by JimBobSquarePants** [reply]
PR created. https://github.com/SixLabors/ImageSharp/pull/1797

**Reply by JimBobSquarePants** [reply]
I'd rather go "good enough" than bloat out the `PixelOperations` API but we'll see. Using `RgbaVector` is an improvement anyway and everything we do is single precision so not even reliable machine to machine.

Plus...
> Digging deeper, I see that values are converted to single floating point values under the hood; this still works for me.

**Reply by calrsom** [reply]
Sorry I've been out of the loop here, but I did want to update the discussion with my use case.

I am rendering agricultural crop and product coverage maps.  I have a custom pixel type that stores the data as floating point values; this is convenient, since the data values vary considerably in magnitude based on the underlying data measurement units.  I initially used the add blend mode to calculate overlapping coverage.  I have since written a custom brush, because I discovered I needed more control over the pixel blending math.

As a point of interest, here is the custom brush I came up with:
```
    /// <summary>
    /// Custom brush for compositing coverage data
    /// </summary>
    public class DataBrush : IBrush
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="DataBrush"/> class.
        /// </summary>
        /// <param name="value">Data values to composite</param>
        public DataBrush(TaPixel value) => Data = value;

        /// <summary>
        /// Gets the color.
        /// </summary>
        public TaPixel Data { get; }

        /// <inheritdoc />
        public BrushApplicator<TPixel> CreateApplicator<TPixel>(
            Configuration configuration,
            GraphicsOptions options,
            ImageFrame<TPixel> source,
            RectangleF region)
            where TPixel : unmanaged, IPixel<TPixel>
        {
            if (source is ImageFrame<TaPixel> dataSource)
            {
                return new DataBrushApplicator(configuration, options, dataSource, Data) as BrushApplicator<TPixel>;
            }
            throw new NotSupportedException("DataBrush only works on TaPixel images");
        }

        /// <summary>
        /// The data brush applicator.
        /// </summary>
        private sealed class DataBrushApplicator : BrushApplicator<TaPixel>
        {
            private TaPixel Data;

            /// <summary>
            /// Initializes a new instance of the <see cref="DataBrushApplicator"/> class.
            /// </summary>
            /// <param name="configuration">Ignored.</param>
            /// <param name="options">Ignored.</param>
            /// <param name="source">The source image.</param>
            /// <param name="data">Data values to composite</param>
            public DataBrushApplicator(
                Configuration configuration,
                GraphicsOptions options,
                ImageFrame<TaPixel> source,
                TaPixel data)
                : base(configuration, options, source)
            {
                Data = data;
            }

            /// <inheritdoc />
            public override void Apply(Span<float> scanline, int x, int y)
            {
                Span<TaPixel> destinationRow = this.Target.GetPixelRowSpan(y).Slice(x);

                // Constrain the spans to each other
                if (destinationRow.Length > scanline.Length)
                    destinationRow = destinationRow.Slice(0, scanline.Length);

                for (int i = 0; i < destinationRow.Length; i++)
                {
                    // Target add blend capped, Actual add blend; see https://en.wikipedia.org/wiki/Blend_modes
                    destinationRow[i] = new TaPixel(
                        Math.Min(destinationRow[i].Target + Data.Target * scanline[i], Math.Max(destinationRow[i].Target, Data.Target)), 
                        destinationRow[i].Actual + Data.Actual * scanline[i]);
                }
            }
        }
    }
```

**Reply by antonfirsov** [reply]
@calrsom thanks! What is the content of `TaPixel`? Two `double`s?

Note that we added support for backing `Color` by an arbitrary `IPixel<T>` implementation in https://github.com/SixLabors/ImageSharp/pull/1801. You should be able to create it with `Color.FromPixel(new TaPixel(...))`.

**Edit:** I see you already found the PR 👍
